### PR TITLE
Port GATK 3 PR 1389

### PR DESCRIPTION
**Pull request by davidbenjamin**
This closes issue #1950. 

@vruano should probably be a part of this as the author of broadinstitute/gsa-unstable#1389, but @droazen surely has opinions as to when to merge these changes.

### CODE REVIEW

**Comment by droazen** [comment]
@davidbenjamin Can you tell me whether this is a straight-up port of the GATK3 version of this code, or whether you've made any changes in the process of porting?

I will test out this change, in combination with a change from @samuelklee / @ronlevine, in a branch in protected, and craft passing integration tests there before merging this here in public.

**Comment by davidbenjamin** [comment]
@droazen This is a straight-up port -- it was copy and paste except I had to update some method names, like `alleleCount()` -> `numberOfAlleles()` and `sampleIndex` -> `indexOfSample`.  And of course I couldn't port the changes to integration tests md5s.

**Comment by vruano** [comment]
Done with my review.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerGenotypingEngine.java [hunk 1]
```diff
@@ -200,6 +201,10 @@ public CalledHaplotypes(final List<VariantContext> calls, final Set<Haplotype> c
             }
         }
 
+        if (call.getAlleles().size() != readAlleleLikelihoodsForAnnotations.numberOfAlleles()) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerGenotypingEngine.java]: this isn't tested

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/ReadLikelihoods.java [hunk 2]
```diff
@@ -881,42 +887,62 @@ public void addNonReferenceAllele(final A nonRefAllele) {
         Utils.nonNull(nonRefAllele, "non-ref allele cannot be null");
         if (!nonRefAllele.equals(GATKVCFConstants.NON_REF_SYMBOLIC_ALLELE)) {
             throw new IllegalArgumentException("the non-ref allele is not valid");
-        }
-        if (alleles.containsAllele(nonRefAllele)) {
+        } else if (alleles.containsAllele(nonRefAllele)) {
             return;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/ReadLikelihoods.java]: neither is this

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/ReadLikelihoods.java [hunk 3]
```diff
@@ -881,42 +887,62 @@ public void addNonReferenceAllele(final A nonRefAllele) {
...
+     * </p>
+     *
+     *
+     * @param allelesToConsider
+     */
+    @SuppressWarnings("unchecked")  // for the cast (A) GATKVCFConstants.NON_REF_SYMBOLIC_ALLELE below
+    public void updateNonRefAlleleLikelihoods(final AlleleList<A> allelesToConsider) {
+        final int alleleCount = alleles.numberOfAlleles();
+        final int nonRefAlleleIndex = indexOfAllele((A) GATKVCFConstants.NON_REF_SYMBOLIC_ALLELE);
+        final int nonSymbolicAlleleCount = nonRefAlleleIndex < 0 ? alleleCount : alleleCount - 1;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/ReadLikelihoods.java]: only one branch is tested here

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/ReadLikelihoods.java [hunk 4]
```diff
@@ -881,42 +887,62 @@ public void addNonReferenceAllele(final A nonRefAllele) {
...
+        for (int s = 0; s < samples.numberOfSamples(); s++) {
+            final double[][] sampleValues = valuesBySampleIndex[s];
+            final int readCount = sampleValues[0].length;
+            for (int r = 0; r < readCount; r++) {
+                final BestAllele bestAllele = searchBestAllele(s, r, true);
+                int numberOfQualifiedAlleleLikelihoods = 0;
+                for (int i = 0; i < alleleCount; i++) {
+                    final double alleleLikelihood = sampleValues[i][r];
+                    if (i != nonRefAlleleIndex && alleleLikelihood < bestAllele.likelihood
+                            && !Double.isNaN(alleleLikelihood) && allelesToConsider.indexOfAllele(alleles.getAllele(i)) != -1) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/ReadLikelihoods.java]: same here

File: src/main/java/org/broadinstitute/hellbender/utils/genotyper/ReadLikelihoods.java [hunk 5]
```diff
@@ -881,42 +887,62 @@ public void addNonReferenceAllele(final A nonRefAllele) {
...
+                        qualifiedAlleleLikelihoods[numberOfQualifiedAlleleLikelihoods++] = alleleLikelihood;
+                    }
+                }
+                final double nonRefLikelihood = medianCalculator.evaluate(qualifiedAlleleLikelihoods, 0, numberOfQualifiedAlleleLikelihoods);
+                // when the median is NaN that means that all applicable likekihoods are the same as the best
+                // so the read is not informative at all given the existing alleles. Unless there is only one (or zero) concrete
+                // alleles with give the same (the best) likelihood to the NON-REF. When there is only one (or zero) concrete
+                // alleles we set the NON-REF likelihood to NaN.
+                sampleValues[nonRefAlleleIndex][r] = !Double.isNaN(nonRefLikelihood) ? nonRefLikelihood
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/genotyper/ReadLikelihoods.java]: same here

**Comment by droazen** [comment]
The source of the index out-of-bounds errors, as @lbergelson discovered, was a line or two of code that was not ported from @vruano 's original PR (https://github.com/broadinstitute/gsa-unstable/pull/1389)

**Comment by droazen** [comment]
Closing in favor of https://github.com/broadinstitute/gatk/pull/3519, which includes this fix.

**Comment by davidbenjamin** [comment]
Great!

**Pull request closed without merging**